EIPS/EIP-X-metadata-nickjohnson.md

```diff
@@ -0,0 +1,72 @@
+## Preamble
+
+    EIP: <to be assigned>
+    Title: Address metadata registry
+    Author: Nick Johnson <nick@ethereum.org>
+    Type: Standard track
+    Category: ERC
+    Status: Draft
+    Created: 2018-03-12
+
+## Abstract
+This EIP specifies a registry for address metadata, permitting both contracts and external accounts to supply metadata about themselves to onchain and offchain callers. This permits use-cases such as generalised authorisations, providing token acceptance settings, and claims registries.
+
+## Motivation
+An increasing set of use cases require storage of metadata associated with an address; see for instance EIP 777 and EIP 780, and the ENS reverse registry in EIP 181. Presently each use-case defines its own specialised registry. To prevent a proliferation of special-purpose registry contracts, we instead propose a single standardised registry using an extendable architecture that allows future standards to implement their own metadata standards.
+
+## Specification
+The metadata registry has the following interface:
+```
+interface AddressMetadataRegistry {
+  function provider(address target) view returns(address);
+  function setProvider(address _provider);
+}
+```
+
+`setProvider` specifies the metadata registry to be associated with the caller's address, while `provider` returns the address of the metadata registry for the supplied address.
+
+The metadata registry will be compiled with an agreed-upon version of Solidity and deployed using the trustless deployment mechanism to a fixed address that can be replicated across all chains.
+
+## Provider specification
+
+Providers may implement any subset of the metadata record types specified here. Where a record types specification requires a provider to provide multiple functions, the provider MUST implement either all or none of them. Providers MUST throw if called with an unsupported function ID.
+
+Providers have one mandatory function:
+
+```
+function supportsInterface(bytes4 interfaceID) constant returns (bool)
+```
+
+The `supportsInterface` function is documented in [EIP 165](https://github.com/ethereum/EIPs/issues/165), and returns true if the provider implements the interface specified by the provided 4 byte identifier. An interface identifier consists of the XOR of the function signature hashes of the functions provided by that interface; in the degenerate case of single-function interfaces, it is simply equal to the signature hash of that function. If a provider returns `true` for `supportsInterface()`, it must implement the functions specified in that interface.
+
+`supportsInterface` must always return true for `0x01ffc9a7`, which is the interface ID of `supportsInterface` itself.
```

Comment:
What is the point of this?  At first blush it feels like it is just a clever little thing to chuckle at in the spec, but not something that anyone would ever need for anything.  I'm guessing the thought is that you can call this to check if `supportsInterface` is available at all (with the expectation that it will return false by default) but I believe there is currently a "bug" in Solidity that makes the default return garbage (which evaluates to false 1/2^256 times).  Even if this is "fixed" (unclear if that is planned or not), it feels weird probing a contract like this.

Is there some other reason for this `MUST`?

Reply:
Also, recommend changing `must` to `**MUST**`.

Reply:
Also see 165 and 137; the idea is that it lets a generic caller determine if the contract implements `supportsInterface` in the first place.

Reply:
As mentioned over in #165 (and I didn't see a rebuttal, though I may have missed it) older Solidity (based on discussion there it sounds like this has since been fixed) contracts will return garbage data by default (meaning true or false).  Thus if someone calls `supportsInterface` on such a contract they may (likely) get back `true` even though it doesn't actually support the interface.  One recommendation was to have both a positive and a negative check function to address this.

Reply:
It appears that 165 addresses this problem with:
```
true when interfaceID is 0x01ffc9a7 (EIP165 interface)
false when interfaceID is 0xffffffff
```

Reply:
> As mentioned over in #165 (and I didn't see a rebuttal, though I may have missed it) older Solidity (based on discussion there it sounds like this has since been fixed) contracts will return garbage data by default (meaning true or false).

Only if they implement a fallback that doesn't throw - which this standard explicitly prohibits.